router_test.go

```diff
@@ -0,0 +1,188 @@
+package main
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestRedirect(t *testing.T) {
+	t.Run("invalid", func(t *testing.T) {
+		s := NewServer()
+
+		req := httptest.NewRequest("GET", "/", nil)
+		w := httptest.NewRecorder()
+		s.redirect(w, req)
+
+		resp := w.Result()
+		body, _ := ioutil.ReadAll(resp.Body)
+
+		if http.StatusBadRequest != resp.StatusCode {
+			t.Fail()
```

Comment:
same as before: you probably should use t.Error instead of t.Fail and add some context.